sympy/ntheory/primetest.py

```diff
@@ -713,8 +740,16 @@ def isprime(n):
     #    https://oeis.org/A014233
     if n < 341531:
         return mr(n, [9345883071009581737])
-    if n < 885594169:
-        return mr(n, [725270293939359937, 3569819667048198375])
+    if n < 4296595241:
+        # Michal Forisek and Jakub Jancina,
+        # Fast Primality Testing for Integers That Fit into a Machine Word
+        # https://ceur-ws.org/Vol-1326/020-Forisek.pdf
+        h = ((n >> 16) ^ n) * 0x45d9f3b
+        h = ((h >> 16) ^ h) * 0x45d9f3b
+        h = ((h >> 16) ^ h) & 255
+        return mr(n, [_MR_BASES_32[h]])
+    # if n < 885594169:
```

Comment:
testing I did suggests that up to 885... those lines of code are about 5% faster then the `n<429...` case. Can they just be kept above the added lines, but after `n<341531` case?

Reply:
How did you measure it?  
In my measurement, it runs faster than the code using two bases.

```python
from sympy import *
from sympy.ntheory.primetest import mr
import random

ns = []
for _ in range(10000):
    n = random.randrange(341531, 885594169, 2)
    if min(factorint(n).keys()) <= 47:
        continue
    ns.append(n)
```

```python
%%timeit
for n in ns:
    mr(n, [725270293939359937, 3569819667048198375])
```

```
24.6 ms ± 171 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

```python
%%timeit
for n in ns:
    h = ((n >> 16) ^ n) * 0x45d9f3b
    h = ((h >> 16) ^ h) * 0x45d9f3b
    h = ((h >> 16) ^ h) & 255
    mr(n, [_MR_BASES_32[h]])
```

```
23 ms ± 184 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
I tested `mr` directly by passing `i` to it instead of to `factorint`.

Reply:
First, the number of times `mr` is executed varies depending on whether `n` is a prime number. In other words, the proportion of primes among the numbers tested affects the speed.

|           | mr1    | mr2      |
|-----------|--------|----------|
| prime     |   1    |   2      |
| composite |   1    | 1 or 2   |

The range `range(5, 885594169, 1000)` consists entirely of multiples of 5 and can be determined using `mr(n, [725270293939359937])`. Therefore, the number of `mr` executions is reduced to just one, and `mr1` becomes slower only by the time required for hash calculations.  
For the range `range(7, 885594169, 1000)`, there are approximately 110,000 prime numbers, but all composites except for one value of `n` can also be determined with `mr(n, [725270293939359937])`. On the other hand, composites with prime factors of 47 or less have already been identified through trial division, so the number of composites that might reach this code block is even smaller. In practice, testing with `ns = [n for n in range(7, 885594169, 1000) if min(factorint(n).keys()) > 47]` shows that the difference between `mr1` and `mr2` narrows.

What surprised me was the significant amount of time spent on hash calculations. This could potentially be improved by using a different function ([e.g.](https://www.techneon.com/download/is.prime.32.base.data), `h = ((0x43fbf0c5 * n) >> 24) % 256`).

As described above, it does not appear to be that slow if actual use is assumed.

Reply:
My reasoning was to send the range of numbers that would be intercepted by the commented out code block. Technically, I should also increase the lower bound to skip over the 341531 bound of the first check. So each code block was tested to compare how the speed to do the steps before calling mr (and then calling mr) compared. I wasn't looking at what numbers might actually get to this code from `factorint`. I trust your code sense in ntheory. Let's go with this and if you want to change the function, that can be a new PR.

Reply:
I would move that commented code to the comment of the PR, however. There's no advantage to keeping them in the code base once this PR is committed.

Reply:
Thank you for your review. I apologize for creating the pull request that doesn’t clearly improve the performance.
I will move the code that was commented out.

Reply:
I did not mean to imply that anything is wrong with the PR. Even if it is a neutral change in timing for numbers less than the original smaller limit, it still extends that limit, and is worthwhile.